src/functions_math.jl

```diff
@@ -83,6 +83,8 @@ macro declare_matmul(MatrixT, VectorT=nothing)
 end
 
 @declare_matmul(AbstractMatrix, AbstractVector)
+@declare_matmul(Adjoint{<:Any, <:AbstractMatrix{T1}} where T1, Adjoint{<:Any, <:AbstractVector})
+@declare_matmul(Diagonal, AbstractVector)
```

Comment:
@oxinabox ideas here?

Adding `@declare_matmul(Diagonal, AbstractVector)` is kind of awkward, because AbstractVector isn't a 1D version of Diagonal. What happens is that certain methods get overwritten, for example
`*(Vector, 2D NDA)`

However it is needed in order to cover `*(Diagonal, 1D NDA)` case.

Reply:
maybe `@declare_matmul` needs to be modified so that
`@declare_matmul(Diagonal)` does the right thing.

Reply:
Hmm, ok, that's probably beyond my current knowledge of Julia. I could hack something but it wouldn't be pretty.
Maybe we should just remove L87 and open a new issue?

Reply:
Can you just write out the methods you want to create directly and explictly?

That will solve it, and maybe later we can refactor it to be done in the macro.

Reply:
I've added the two methods manually. Once this PR is merged I will open an issue highlighting this.